src/System.Diagnostics.FileVersionInfo/src/System/Diagnostics/FileVersionInfo.Unix.cs

```diff
@@ -9,42 +14,195 @@ private FileVersionInfo(string fileName)
+                                _fileMinor = v.Minor;
+                                _fileBuild = v.Build;
+                                _filePrivate = v.Revision;
+
+                                // When the managed compiler sees an [AssemblyVersion(...)] attribute, it uses that to set 
+                                // both the assembly version and the product version in the Win32 resources. If it doesn't 
+                                // see an [AssemblyVersion(...)], then it sets the assembly version to 0.0.0.0, however it 
+                                // sets the product version in the Win32 resources to whatever was defined in the 
+                                // [AssemblyFileVersionAttribute(...)] if there was one.  Without parsing the Win32 resources,
+                                // we can't differentiate these two cases, so given the rarity of explicitly setting an 
+                                // assembly's version number to 0.0.0.0, we assume that if it is 0.0.0.0 then the attribute 
+                                // wasn't specified and we use the file version.
+                                if (_productVersion == "0.0.0.0")
+                                {
+                                    _productVersion = _fileVersion;
+                                    _productMajor = _fileMajor;
+                                    _productMinor = _fileMinor;
+                                    _productBuild = _fileBuild;
+                                    _productPrivate = _filePrivate;
+                                }
+                            }
+                            break;
+                        case "System.Reflection.AssemblyProductAttribute":
+                            _productName = attrNameAndValue.Value;
+                            break;
+                        case "System.Reflection.AssemblyTrademarkAttribute":
+                            _legalTrademarks = attrNameAndValue.Value;
+                            break;
+                        case "System.Reflection.AssemblyTitleAttribute":
+                            _fileDescription = attrNameAndValue.Value;
+                            break;
+                    }
+                }
+            }
         }
 
-        // -----------------------------
-        // ---- PAL layer ends here ----
-        // -----------------------------
+        /// <summary>
+        /// Attempt to parse a CustomAttribute into its full name and the value of its single string parameter.
+        /// </summary>
+        /// <param name="reader">The metadata reader.</param>
+        /// <param name="attr">The attribute to parse.</param>
+        /// <param name="result">The name of the attribute and its parameter's value, if it could be parsed.</param>
+        /// <returns>true if the attribute could be parsed; otherwise, false.</returns>
+        private static bool TryGetAttributeNameAndStringArgument(
```

Comment:
You probably also want to catch (BadImageFormatException) and return false too. Some errors, such as hitting the end of an invalid attribute blob would manifest that way, while others (bad prolog) you're handling by returning false. I think it's reasonable to just skip invalid metadata here.